Mathlib/Geometry/Manifold/Orientable.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2024 Rida Hamadani. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rida Hamadani
+-/
+import Mathlib.Analysis.Calculus.FDeriv.Comp
+import Mathlib.Geometry.Manifold.ChartedSpace
+import Mathlib.Topology.Algebra.Module.Determinant
+
+/-!
+# Orientable Manifolds
+
+This module defines orientable manifolds.
+
+## Main Definitions
+
+* `OrientationPreserving` : a map is orientation-preserving on a given set if the determinant of
+  its Jacobian is strictly positive on that set.
+* `orientationPreservingGroupoid` : the groupoid of partial homeos of `H` which are
+  orientation-preserving.
+* `OrientableManifold`: a type class saying that the charted space `M`, modelled on the space `H`,
+  admis an orientation. This type class is just a shortcut for `HasGroupoid M
+  (@orientationPreservingGroupoid H _ _)`.
+
+-/
+
+variable {H : Type*} [NormedAddCommGroup H] [NormedSpace ℝ H]
+
+section orientationPreservingGroupoid
+
+/--
+A map is orientation-preserving on a given set if it is differentiable and the determinant of its
+Jacobian is strictly positive on that set.
+-/
+def OrientationPreserving (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s,
```

Comment:
Also: I think you can remove the differentiability requirement, as `fderiv` has junk value `0` whenever the map is not differentiable. (Perhaps add a comment to that effect, if you find one helpful.)
That ought to simplify later proof, right?

Reply:
I've considered doing that, and it turns out that I will need to add another condition to get differentiability from the positivity of `fderiv`, one possible condition to obtain that is `FiniteDimensional.finrank ℝ H ≠ 0` (cf. `LinearMap.det_zero`). Since I need to add a condition anyways, I decided to go with the one that sounded to me closer to spoken mathematics. Do you think that the `FiniteDimensional.finrank ℝ H ≠ 0` condition is better in our case?